### ISSUE TITLE
MQTT 5 Support with shared subscriptions

### ISSUE THREAD

**Issue by abnerjacobsen** [absmach/magistrala#1235]
I have a question: Does Mainflux support the MQTT 5 protocol, specifically shared subscriptions ? As, for example, it is explained in this document: http://www.steves-internet-guide.com/mqttv5-shared-subscriptions/

Thanks,

**Comment by drasko** [comment]
Mainflux uses [mProxy](https://github.com/mainflux/mproxy) in front of the off-the-shelf MQTT broker of your choice. Default one is [VerneMQ](https://github.com/vernemq/vernemq), and this one does support shared subscriptions: https://docs.vernemq.com/configuration/shared_subscriptions

However, I am not sure if mProxy will let all the topics pass throuigh, as Mainflux uses `channels`, which have a form of topics with defined format of `channels/<channel_id>/message`.

@blokovi @dusanb94 opinions?

**Comment by abnerjacobsen** [comment]
@drasko I did a quick search in the Mainflux source code and found several places in it that have some kind of regex like in line 25 of mqtt/handler.go:

`channelRegExp         = regexp.MustCompile(`^\/?channels\/([\w\-]+)\/messages(\/[^?]*)?(\?.*)?$`)`

I understand that it is not possible to use shared channels with Mainflux, correct?

I would like to suggest support for shared channels as a new feature.

Thank you.

**Comment by drasko** [comment]
![FB_IMG_1600717214729.jpg](https://user-images.githubusercontent.com/149944/93813276-2024f700-fc53-11ea-8d4b-be9905b1edbb.jpg)

**Comment by drasko** [comment]
> @drasko I did a quick search in the Mainflux source code and found several places in it that have some kind of regex like in line 25 of mqtt/handler.go:
> 
> `channelRegExp         = regexp.MustCompile(`^\/?channels\/([\w\-]+)\/messages(\/[^?]*)?(\?.*)?$`)`
> 
> I understand that it is not possible to use shared channels with Mainflux, correct?

Yep, this is what I was afraid of, we have a topic validation, protecting that only API specified endpoints can be allowed.

This shod be changed, and it should not be very difficult, as a broker behind mProxy already supports shared connections.

> I would like to suggest support for shared channels as a new feature.

I agree. Let's add this enhancement.

We'll take a look at this as soon as we have some bandwidth, in a few days.
